pkg/codegen/go/gen.go

```diff
@@ -1878,6 +1880,111 @@ func init() {
 	fmt.Fprintf(w, initCode)
 }
 
+type objectProperty struct {
+	object   *schema.ObjectType
+	property *schema.Property
+}
+
+// When computing the type name for a field of an object type, we must ensure that we do not generate invalid recursive
+// struct types. A struct type T contains invalid recursion if the closure of its fields and its struct-typed fields'
+// fields includes a field of type T. A few examples:
+//
+// Directly invalid:
+//
+//     type T struct {
+//         Invalid T
+//     }
+//
+// Indirectly invalid:
+//
+//     type T struct {
+//         Invalid S
+//     }
+//
+//     type S struct {
+//         Invalid T
+//     }
+//
+// In order to avoid generating invalid struct types, we replace all references to types involved in a cyclical
+// definition with *T. The examples above therefore become:
+//
+// (1)
+//     type T struct {
+//         Valid *T
+//     }
+//
+// (2)
+//     type T struct {
+//         Valid *S
+//     }
+//
+//     type S struct {
+//         Valid *T
+//     }
+//
+// We do this using a rewriter that turns all fields involved in reference cycles into optional fields.
+func rewriteCyclicField(rewritten codegen.Set, path []objectProperty, op objectProperty) {
+	// If this property refers to an Input<> type, unwrap the type. This ensures that the plain and input shapes of an
+	// object type remain identical.
+	t := op.property.Type
+	if inputType, isInputType := op.property.Type.(*schema.InputType); isInputType {
+		t = inputType.ElementType
+	}
+
+	// If this property does not refer to an object type, it cannot be involved in a cycle. Skip it.
+	objectType, isObjectType := t.(*schema.ObjectType)
+	if !isObjectType {
+		return
+	}
+
+	path = append(path, op)
+
+	// Check the current path for cycles by crawling backwards until reaching the start of the path
+	// or finding a property that is a member of the current object type.
+	var cycle []objectProperty
+	for i := len(path) - 1; i > 0; i-- {
+		if path[i].object == objectType {
```

Comment:
Correct to use pointer equality here? one physical object per logical type in the process?

Reply:
For object types, yes. The schema binder ensures that reference equality indicates type identity for these types.